### DISCUSSION TITLE
Allow to disable conditionally

### DISCUSSION THREAD

**Discussion by ouvreboite** [picocss/pico#604] [Ideas]
Hi,

The conditional version is nice, but it does not solve one of the common problem: using a third-party tool **inside** a pico component. For example, displaying a GoogleMap inside a card or using agGrid inside a section.

Having a `.no-pico` (or `.picoff`) class that would stop the pico styles on the element and all its children would be very useful.

Personally, I would like to push PicoCSS for some of the internal apps where I work, but smoothly integrating with other libraries is a must.

(Maybe there is a simpler solution to this problem ?)

**Comment by dev-willis** [comment]
I've attempted to solve this problem like so:

```scss
@use "@picocss/pico/scss/pico" with (
	$parent-selector: "*:not(#no_pico_inside_this_element *) ",
);
```

This helps but isn't foolproof and feels hacky. I think it might work better if #533 gets resolved.

**Comment by ouvreboite** [comment]
Not yet available in Firefox, but the [@scope](https://developer.mozilla.org/en-US/docs/Web/CSS/@scope) feature could be useful.

**Comment by ouvreboite** [comment]
@dev-willis @lucaslarroche 

I did a small test with [@scope](https://developer.mozilla.org/en-US/docs/Web/CSS/@scope) on Chrome:

- I took the base css file: [pico.css](https://github.com/picocss/pico/blob/main/css/pico.css)
- Wrapped in in a @scope using two custom class as the scope root and limit: `.picon` enables pico while `.picoff` disables it
- ⚠️ [pico.css](https://github.com/picocss/pico/blob/main/css/pico.css) contains a lot of reference to `:root`. This does not work in @scope, but you can simply replace all occurence of `:root` by :`scope`
```css
@charset "UTF-8";
@scope (.picon) to (.picoff) {
  /* the base content of pico.css with all :root replaced by :scope*/
}
```

And voilà 🎉 

![image](https://github.com/user-attachments/assets/daede9e2-7b33-49f7-a2dd-b5dbcd5a6041)

You can aslo re-enable pico inside a disabled block:
```html
<div class="picon">
  this is pico
  <div class="picoff">
    this is not pico
    <div class="picon">
      this is pico again
  </div>
</div>
</div>
```

⚠️ But this seems to increase the font size each time 😨 .
I suppose it's because some of the css variable use %, and the on-and-off-again behavior makes it coumpount.
(`--pico-font-size: 118.75%`)

![image](https://github.com/user-attachments/assets/941a3b3c-100b-46cb-9a84-5a6e26373775)

**Comment by Sleepful** [comment]
This feature is gravely missing from Pico, this would allow Pico to be much more, by allowing it to be a base that does not get in the way. Right now it gets in the way and it is very difficult to undo the styling from, say, a `<button>`. Notice I do want the `button` to be styled most of the time, but some of the time I need it to look differently, like a link with a lot less style.

**Comment by Sleepful** [comment]
For now I solved it with a Shadow Root. 
Do take into account that using a Shadow Root can get quite tricky depending on the JavaScript framework that you are using.
It solves the problem perfectly though, as it is meant to provide style scoping.

Here is how I did it with SolidJS:

https://github.com/solidjs/solid/issues/1834#issuecomment-2951922846

**Comment by ricpastori** [comment]
The integration problem is real, but I don’t think a `.no-pico` selector can reliably switch a global stylesheet off.

Excluding a subtree from every selector does not undo inherited values, custom properties or resets that have already crossed the boundary. It would also multiply the complexity and specificity of nearly every rule.

A scoped build that only activates inside `.pico`, or isolating the third-party widget with Shadow DOM, is more predictable. In the future, `@scope` with a scope limit may provide a cleaner native solution once support matches Pico’s browser policy.

**Reply by Leftium** [reply]
My nimble.css uses `@scope` to conditionally disable styles (and looks/works very similar to PicoCSS)

Demo: https://leftium.github.io/nimble.css/demo/extended.html#no-nimble

**Reply by ricpastori** [reply]
Thanks, this is a useful counterexample. Please correct me if I have misunderstood how nimble.css works.

If I understand the implementation correctly, `@scope` provides a reliable boundary for the component-level selectors, while *nimble.css* deliberately keeps its reset, custom properties, document styles, layout utilities, and print styles global. Inherited properties such as `font-family` and `color` can also continue across the scope limit.

So `.no-nimble` does not provide complete stylesheet isolation in the same sense as Shadow DOM. It offers a deliberately narrower and very useful contract: component styles stop applying, while the shared baseline remains active.

That distinction changes my assessment somewhat. A `.no-pico` mechanism based on `@scope` could be viable if it explicitly promised this kind of component-style opt-out, rather than guaranteeing that the subtree is entirely unaffected by Pico.

One remaining concern for Pico would be its browser-support policy. If supporting older browsers such as Internet Explorer is still a goal, `@scope` cannot be the only implementation. Treating this as progressive enhancement also seems questionable. In an unsupported or affected browser, the failure mode is that Pico's component styles leak into the third-party widget, which is precisely the problem `.no-pico` is supposed to prevent. This is not as harmless as losing an optional visual enhancement.

I also noticed that nimble.css leaves the prebuilt stylesheet unscoped by default, while the demo uses `no-nimble.js` to apply the scope progressively and to opt out when it detects an affected browser. Is my understanding correct?

**Reply by Leftium** [reply]
That seems like a pretty good assessment.
- `.no-nimble` intentionally allows some global styles to keep working
- `@scope` only works in relatively modern browsers
- nimble.css chooses to support a pragmatic set of modern browsers (i.e. no support for IE Explorer, etc)
- There is a bug in Safari 18 that causes `@scope` to not work in nimble.css. (Styling becomes very broken: worse than just leaking styles)
  - Couldn't find a way around this without JS, so had to implement `.no-nimble` as dynamically injected `@scope` via JS
  - Hopefully after Safari 18 becomes less common, `@scope` can be used directly without JS injection.
  
It's been a long time since I made these design decisions (with coding agent assistance), coding agents may be able to make more accurate assessments.

Nimble.css generally works in my projects (and devices). I fix little bugs as I discover them. So far, it's been a great replacement for PicoCSS: https://leftium.github.io/nimble.css/demo/pico.html

**Reply by Leftium** [reply]
I just reviewed nimble.css `.no-css` and re-implemented it.
- I actually could not reproduce the CSS issues (on Safari 18)
- And the coding agent suggested a better CSS-only implementation without JS
- Examine the recent commits and spec documenting the changes: https://github.com/Leftium/nimble.css/blob/main/specs/no-nimble-compatibility.md

**Reply by ricpastori** [reply]
Thank you, I built both versions rather than take the change on trust, and it corrects my earlier objection.

**I was wrong to group this with `@scope`.** `@scope` reached Baseline only in December 2025,
with Firefox 146, which is why I raised browser support as a blocker. The guard needs only `:where()` and a complex selector inside `:not()` — Chrome 88, Firefox 84, Safari 9.

Pico v2 already ships `:has()` 23 times, which puts its real floor at December 2023 at the earliest, whatever the stated policy. And in every engine complex `:not()` shipped well before `:has()` did. So any browser that can render Pico v2 today can already parse the guard; it cannot be the thing that blocks this. The objection I raised does not apply to your implementation.

Two things I checked against 0.21.3:

- **Specificity is genuinely untouched.** Stripping the guard from the compiled output gives back the previous selector set exactly, apart from three `*` elisions that match identically. That was the part I expected to go wrong.
- **The cost compresses away.** +27% minified, but +3.9% gzipped (5,148 → 5,349 B). The raw number is not what consumers pay.

Behaviour is correct in Chromium 149, WebKit 26.5 and Firefox 151 — current engines, not Safari 18, so I can't speak to the historical report.

One caveat worth documenting: the reset stays global, so inside `.no-nimble` a `div` still computes `box-sizing: border-box`, controls still take `font: inherit` (16px, not the UA's 13.3px), and `prefers-reduced-motion` still applies `animation-duration: 0.01ms !important`. That fully answers @Sleepful's case. For @ouvreboite's agGrid it gets most of the way, and those three are the likely remaining surprises — worth naming in the docs rather than leaving them implied by "Reset".

**Reply by ricpastori** [reply]
Separately, and with credit where it's due: I maintain [Cirth](https://github.com/cirthcss/cirth), another semantic-first framework in this space, and `nimble.css` has been the most useful thing I've read on this problem.

I've opened issues to follow your lead on three things:
- the CSS-only exclusion guard,
- moving the generated stylesheets into cascade layers,
- the habit your `specs/` directory demonstrates: writing the contract and an honest evidence ledger before the implementation, including the parts that could not be verified.

Thank you for working in the open.
